### DISCUSSION TITLE
Add Argument to disable merging of mesh primitives (or making them instances)

### DISCUSSION THREAD

**Discussion by BoyBaykiller** [zeux/meshoptimizer#678] [Ideas]
Wether mesh primitives are merged has multiple implications for my (and other peoples) renderer.
Most obviously merged primitives share per mesh(-instance) properties like transformations or additional material data affecting individual control.

Picture the curtains in Sponza. I think it makes sense to have each curtain be its own distinct mesh primitive, instead of having ones with same material be merged. Especially because they are disconnected which can affect spatial data structures. After running gltfpack over the [Sponza model](https://github.com/KhronosGroup/glTF-Sample-Models/tree/main/2.0/Sponza) this is no longer the case (no matter if I use `-kn -km`).

Baiscally I am interested, [just like others have been](https://github.com/zeux/meshoptimizer/issues/470), in an argument that disables this:
https://github.com/zeux/meshoptimizer/blob/260508938a45038f421b29a94c1135fe312a5d7e/gltf/gltfpack.cpp#L364

---

It would also be interesting to have otherwise merged mesh primitives become instances instead (using `EXT_mesh_gpu_instancing`).  Then the engine could still be aware of them being distinct entities, solving the aforementioned problems, while also ensuring efficient rendering. But this is just an idea and and most likely out of scope.

**Comment by zeux** [comment]
> merged primitives share per mesh(-instance) properties like transformations or additional material data affecting individual control

That's why `-kn` flag exists though. The scene your linked has unnamed nodes; I'm not sure how you were planning to identify and individually control them? Re: instancing, this is supported via `-mi` flag but it of course requires identical geometry, for Sponza I assume the geometry of different curtains is not the same.

**Comment by BoyBaykiller** [comment]
I did try the `-kn` flag:
```
input: 103 mesh primitives (262267 triangles, 192496 vertices); 103 draw calls (103 instances, 262267 triangles)
output: 25 mesh primitives (262267 triangles, 192496 vertices); 25 draw calls (25 instances, 262267 triangles)
```
It is merging mesh primitives. What I mean is that I can't individually control each of the original 103 mesh primitives if they get merged into 25. Obviously because then the model loader only creates 25 meshes. And there are other reasons for why you might not want them to be merged.
